subsys/usb/class/mass_storage.c

```diff
@@ -321,13 +321,22 @@ static bool requestSense(void)
 
 static bool inquiryRequest(void)
 {
-	uint8_t inquiry[] = { 0x00, 0x80, 0x00, 0x01,
-	36 - 4, 0x80, 0x00, 0x00,
-	'Z', 'E', 'P', 'H', 'Y', 'R', ' ', ' ',
-	'Z', 'E', 'P', 'H', 'Y', 'R', ' ', 'U', 'S', 'B', ' ',
-	'D', 'I', 'S', 'K', ' ',
-	'0', '.', '0', '1',
-	};
+	const char *vendor_id = CONFIG_MASS_STORAGE_INQ_VENDOR_ID;
+	const char *product_id = CONFIG_MASS_STORAGE_INQ_PRODUCT_ID;
+	const char *revision = CONFIG_MASS_STORAGE_INQ_REVISION;
+
+	uint8_t inquiry[36] = {0x00, 0x80, 0x00, 0x01,
+				36 - 4, 0x80, 0x00, 0x00};
+	/* vendor id is max 8 characters, product id max 16,
+	 * and revision max 4. pad with spaces.
+	 */
+	memset(&inquiry[8], ' ', ARRAY_SIZE(inquiry) - 8);
+
+	memcpy(&inquiry[8], vendor_id, MIN(strlen(vendor_id), 8));
+
+	memcpy(&inquiry[16], product_id, MIN(strlen(product_id), 16));
+
+	memcpy(&inquiry[32], revision, MIN(strlen(revision), 4));
```

Comment:
I understand the motivations but it does not make the code better. I suggest alternative, this is derived from https://github.com/zephyrproject-rtos/zephyr/pull/23483 (I started to rework the SCSI subset used by MSC, but postponed it because of stack rework and it is probably better to rewrite MSC implementation completely.):

```diff c
diff --git a/subsys/usb/class/mass_storage.c b/subsys/usb/class/mass_storage.c
index 4739847b4d..eab34cd024 100644
--- a/subsys/usb/class/mass_storage.c
+++ b/subsys/usb/class/mass_storage.c
@@ -319,17 +319,23 @@ static bool requestSense(void)
        return write(request_sense, sizeof(request_sense));
 }
 
+struct dabc_inquiry_data {
+       uint8_t head[8];
+       uint8_t t10_vid[8];
+       uint8_t product_id[16];
+       uint8_t product_rev[4];
+} __packed;
+
+static const struct dabc_inquiry_data inq_rsp = {
+       .head = {0x00, 0x80, 0x00, 0x01, 36 - 4, 0x80, 0x00, 0x00},
+       .t10_vid = CONFIG_MASS_STORAGE_INQ_VENDOR_ID,
+       .product_id = CONFIG_MASS_STORAGE_INQ_PRODUCT_ID,
+       .product_rev = CONFIG_MASS_STORAGE_INQ_REVISION,
+};
+
 static bool inquiryRequest(void)
 {
-       uint8_t inquiry[] = { 0x00, 0x80, 0x00, 0x01,
-       36 - 4, 0x80, 0x00, 0x00,
-       'Z', 'E', 'P', 'H', 'Y', 'R', ' ', ' ',
-       'Z', 'E', 'P', 'H', 'Y', 'R', ' ', 'U', 'S', 'B', ' ',
-       'D', 'I', 'S', 'K', ' ',
-       '0', '.', '0', '1',
-       };
-
-       return write(inquiry, sizeof(inquiry));
+       return write((uint8_t *)&inq_rsp, sizeof(inq_rsp));
 }
```

Reply:
that looks like a more thorough fix - what's the plan with that PR? is it ever going to get merged?

we're using mass storage in real products, where we obviously want to change these strings. currently we have local changes to the `inquiry` array.

Reply:
>that looks like a more thorough fix - what's the plan with that PR? is it ever going to get merged?

It will probably take a long time :-)

>or should i bring over some of the changes here?

You can pull the patch from [here](https://github.com/jfischer-no/zephyr/commit/e8b331d5bf3d9fae9f30aa30ae9127d57275cbed.patch), just squash it to your commit.

Reply:
right. however, there is one difference from the struct version compared to mine/the previous: the strings will be NULL terminated. i don't have a way of testing this at the moment, so i'm not sure whether that is an issue or not.

Reply:
> right. however, there is one difference from the struct version compared to mine/the previous: the strings will be NULL terminated. i don't have a way of testing this at the moment, so i'm not sure whether that is an issue or not.

I had it first with memcpy(), but I think it is not necessary here.
`uint8_t t10_vid[] = "zephyr"; /* 'z','e','p','h','y','r','\n' */`
is not the same as
`uint8_t t10_vid[8] = "zephyr"; /* 'z','e','p','h','y','r' */`

Reply:
a colleague tested the current patch, and it works - *but* according to https://www.seagate.com/files/staticfiles/support/docs/manual/Interface%20manuals/100293068j.pdf (page 28, "Left aligned") the strings should be padded with 0x20.

Reply:
@jfischer-no : any comments on what the pdf states?

Reply:
That is probably true, I will check it in an other spec. Pity, this approach saves a bit of code size. Would padding with spaces in Kconfig variable and mention it in the Kconfig's help be a solution?

Reply:
well, i guess that would work. i'm not sure what's ugliest - my solution or that :) adding spaces in Kconfig is more optimal from a code size (and complexity) perspective at least, but not entirely user friendly.

if i have to choose between the two, i'd vote for the spaces in Kconfig solution.

Reply:
I am agree with this. I will put a dev-review tag on it.

Reply:
i'm not sure i follow - what does dev-review mean? should i do yet another version with updated Kconfig help?

Reply:
>i'm not sure i follow - what does dev-review mean?

https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Group-Meetings#zephyr-dev-meeting

>should i do yet another version with updated Kconfig help?

Please wait with that.